src/libmongoc/src/mongoc/mongoc-bulkwrite.c

```diff
@@ -0,0 +1,1833 @@
+/*
+ * Copyright 2024-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <mongoc-bulkwrite.h>
+
+#include <bson/bson.h>
+#include <common-macros-private.h> // MC_ENABLE_CONVERSION_WARNING_BEGIN
+#include <mongoc/mcd-nsinfo.h>
+#include <mongoc/mongoc-array-private.h>
+#include <mongoc/mongoc-buffer-private.h>
+#include <mongoc/mongoc-client-private.h>
+#include <mongoc/mongoc-client-side-encryption-private.h>
+#include <mongoc/mongoc-error.h>
+#include <mongoc/mongoc-error-private.h> // _mongoc_write_error_handle_labels
+#include <mongoc/mongoc-server-stream-private.h>
+#include <mongoc/mongoc-util-private.h> // _mongoc_iter_document_as_bson
+
+MC_ENABLE_CONVERSION_WARNING_BEGIN
+
+typedef struct {
+   bool isset;
+   bool val;
+} mongoc_opt_bool_t;
+
+struct _mongoc_bulkwriteopts_t {
+   mongoc_opt_bool_t ordered;
+   mongoc_opt_bool_t bypassdocumentvalidation;
+   bson_t *let;
+   mongoc_write_concern_t *writeconcern;
+   mongoc_opt_bool_t verboseresults;
+   bson_t *comment;
+   bson_t *extra;
+   uint32_t serverid;
+};
```

Comment:
> Wouldn't these typically be defined in a mongoc-bulkwrite-private.h header?

I expect there is no need unless these structs are used in other .c files.

> `mongoc_opt_bool_t` also looks like it may be generally useful

This prompted me to rediscover the `mongoc_optional_t`. Removed `mongo_opt_bool_t` and replaced with use of `mongoc_optional_t`.